### PR TITLE
Refactor actions

### PR DESCRIPTION
**Pull request by piponazo**
Another PR of pure refactoring. Note that this one is not pointing to master but to the branch where I moved the Params implementation to a new .cpp file.

Again, I started these changes with the objective of improving the Singleton implementation of TaskFactory (first commit) and then I decided to cleanup more things I found in those files. Since there are many little refactorings, I recommend to review the changes commit by commit (Specially because I applied clang-format to the files touched).

### CODE REVIEW

File: src/actions.cpp
Comment (D4N) [inline src/actions.cpp]: Shouldn't this go into the destructor instead?

File: src/actions.hpp
Comment (D4N) [inline src/actions.hpp]: Any particular reason why you removed this one?

**Review by D4N — CHANGES_REQUESTED** [review]
Overall looks good to me. A few points:

- I'd say we merge this after #784 into master directly (-> please change the target branch).
- There's two "remove useless assert" commits, please squash these together
- The refactoring induced a regression: ASAN detects a memoryleak of TaskFactory. I'd suggest to move `cleanup` into the destructor, which should (in theory) fix the issue.

File: src/actions.cpp
Comment (piponazo) [reply src/actions.cpp]: Right. Let's do that

File: src/actions.hpp
Comment (piponazo) [reply src/actions.hpp]: The only reason is that I do not like typedefs that are not used a lot. In this case I could remove 7-8 lines from the header file by just removing these typedefs. 

It also seems to me a bit unnatural to have 7-8 typedefs called `UniquePtr` in the same file, which are unique pointers to different types.

Comment (D4N) [reply src/actions.hpp]: Makes sense, since this is not a public class, let's drop it then.

**Comment by piponazo** [comment]
> I'd say we merge this after #784 into master directly (-> please change the target branch).
Sure, I'll do it once the other branch is merged.

**Comment by D4N** [comment]
@mergifyio: why on earth is this thing losing random PRs?

**Pull request closed without merging**
